browser/ui/views/playlist/thumbnail_provider.cc

```diff
@@ -0,0 +1,137 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/views/playlist/thumbnail_provider.h"
+
+#include "base/containers/flat_map.h"
+#include "base/containers/lru_cache.h"
+#include "base/files/file_path.h"
+#include "base/files/file_util.h"
+#include "base/no_destructor.h"
+#include "base/task/thread_pool.h"
+#include "base/task/thread_pool/thread_pool_instance.h"
+#include "brave/browser/playlist/playlist_service_factory.h"
+#include "brave/browser/playlist/playlist_tab_helper.h"
+#include "brave/components/playlist/browser/playlist_constants.h"
+#include "brave/components/playlist/browser/playlist_service.h"
+#include "net/base/filename_util.h"
+
+namespace {
+
+using ItemImageCache = base::LRUCache<std::string /*item_id*/, gfx::Image>;
+
+ItemImageCache& GetInMemoryCache(playlist::PlaylistService* service) {
+  auto key = reinterpret_cast<uintptr_t>(service);
+
+  // We don't want to mix up images from different services, as the
+  // PlaylistService is bound to a Profile.
+  static base::NoDestructor<base::flat_map<uintptr_t, ItemImageCache>> s_cache;
+  if (!base::Contains(*s_cache, key)) {
+    s_cache->insert({key, ItemImageCache(/*max size=*/30)});
+  }
+  return s_cache->at(key);
+}
+
+bool IsItemThumbnailCached(const playlist::mojom::PlaylistItemPtr& item) {
+  return item->thumbnail_path.is_valid() &&
+         item->thumbnail_path != item->thumbnail_source;
+}
+
+}  // namespace
+
+ThumbnailProvider::ThumbnailProvider(playlist::PlaylistService* service)
+    : service_(service) {
+  CHECK(service);
+}
+
+ThumbnailProvider::ThumbnailProvider(playlist::PlaylistTabHelper* tab_helper)
+    : ThumbnailProvider(playlist::PlaylistServiceFactory::GetForBrowserContext(
+          tab_helper->web_contents()->GetBrowserContext())) {}
+
+ThumbnailProvider::~ThumbnailProvider() = default;
+
+void ThumbnailProvider::GetThumbnail(
+    const playlist::mojom::PlaylistItemPtr& item,
+    base::OnceCallback<void(const gfx::Image&)> callback) {
+  DVLOG(2) << __FUNCTION__;
+  auto& in_memory_cache = GetInMemoryCache(service_);
+  if (auto iter = in_memory_cache.Get(item->id);
+      iter != in_memory_cache.end()) {
+    std::move(callback).Run(iter->second);
+    return;
+  }
+
+  if (IsItemThumbnailCached(item)) {
```

Comment:
I think we could reduce in-memory cache memory usage.
If item's thumbnail is already cached, we could remove it from in-memory cache?

Reply:
Updated to use in-mem cache only for thumbanils from network. In addition to that, when reading thumbnail from local disk, remove in-mem cache.